{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import h5py, matplotlib\n",
    "import os, sklearn, umap, rpy2\n",
    "import arboresto, kneed\n",
    "import matplotlib_venn\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import pandas as pd\n",
    "import networkx as nx\n",
    "\n",
    "from rpy2 import robjects\n",
    "import rpy2.robjects.numpy2ri\n",
    "robjects.numpy2ri.activate()\n",
    "from rpy2.robjects.packages import importr\n",
    "base = importr(\"base\")\n",
    "dollar = base.__dict__[\"$\"]\n",
    "\n",
    "\n",
    "# custom class for sc datasets\n",
    "# * additional dependencies:\n",
    "#   *\n",
    "\n",
    "from dataset import dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### \n",
    "\n",
    "#-------------------------------------------------#\n",
    "###\n",
    "# label cells by cluster from Lush et al. and genes \n",
    "# by mouse homologues for later dataset integration\n",
    "###\n",
    "fishEns_to_mouseName = pd.read_csv(('metaAnalysis/'\n",
    "                                    'refs/'\n",
    "                                    'fishEns_to'\n",
    "                                    '_mouseName.csv'),\n",
    "                                    index_col=0)\n",
    "\n",
    "labeled_data = pd.DataFrame(data_copy.values,\n",
    "                            index=data_copy.index.values,\n",
    "                            columns=data_copy.columns.values)\n",
    "\n",
    "labeled_data = labeled_data.loc[:,[x for x in \n",
    "                    labeled_data.columns.values if x in \n",
    "                      fishEns_to_mouseName.index.values]]\n",
    "\n",
    "labeled_data.columns = [fishEns_to_mouseName.loc[x,'Associated Gene Name'].values[0] if \n",
    "                        type(fishEns_to_mouseName.loc[x,'Associated Gene Name'])!=str \n",
    "                        else fishEns_to_mouseName.loc[x,'Associated Gene Name'] for x in \n",
    "                        labeled_data.columns.values]\n",
    "\n",
    "labeled_data['node'] = labeled_data.index.to_series(\n",
    "                        ).map(dict(zip(p4i_data.index,\n",
    "                                       p4i_data['node'])))\n",
    "#--------------------------------------------------------#\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###################################################################################################################\n",
    "#######################################  PART 3: META-ANALYSIS   ##################################################\n",
    "###################################################################################################################\n",
    "\n",
    "#-------------------------------------------------------#\n",
    "########## preprocess mouse data for alignment ##########\n",
    "#-------------------------------------------------------#\n",
    "\n",
    "try:\n",
    "    metadata = pd.read_csv(('metaAnalysis/'\n",
    "                            'datasets/'\n",
    "                            'metadata.csv'),\n",
    "                            index_col=0)\n",
    "    \n",
    "    lush = pd.read_csv(('metaAnalysis/'\n",
    "                        'datasets/'\n",
    "                        'lush.csv'),\n",
    "                        index_col=0)\n",
    "    \n",
    "    burns = pd.read_csv(('metaAnalysis/'\n",
    "                         'datasets/'\n",
    "                         'burns.csv'),\n",
    "                         index_col=0)\n",
    "    \n",
    "    hoa = pd.read_csv(('metaAnalysis/'\n",
    "                       'datasets/'\n",
    "                       'hoa.csv'),\n",
    "                       index_col=0)\n",
    "\n",
    "except:  \n",
    "    print('loading datasets...')\n",
    "    \n",
    "    #---------------------------------------#\n",
    "    ### load Burns et al. data (GSE71982) ###\n",
    "    burns_t = pd.read_csv(('GSE71982/'\n",
    "                           'GSE71982_'\n",
    "                           'RSEM_Counts_'\n",
    "                           'Matrix.csv'),\n",
    "                           index_col=0)\n",
    "    \n",
    "    burns_phenoData_utric = pd.read_csv(('GSE71982/'\n",
    "                                         'GSE71982_'\n",
    "                                         'P1_Utricle_'\n",
    "                                         'PhenoData.csv'),\n",
    "                                         index_col=0)\n",
    "    \n",
    "    burns_phenoData_cochl1 = pd.read_csv(('GSE71982/'\n",
    "                                          'GSE71982_'\n",
    "                                          'P1_Coch__'\n",
    "                                          'nonFACs__'\n",
    "                                          'PhenoData.csv'),\n",
    "                                          index_col=0)\n",
    "    \n",
    "    burns_phenoData_cochl2 = pd.read_csv(('GSE71982/'\n",
    "                                          'GSE71982_'\n",
    "                                          'P1_Coch__'\n",
    "                                          'FACs__PhenoData.csv'),\n",
    "                                          index_col=0)\n",
    "    burns = burns_t.T\n",
    "    burns.drop(['EGFP','tdTom'],\n",
    "               axis=1,inplace=True)\n",
    "    #-----------------------------#\n",
    "    \n",
    "    \n",
    "    \n",
    "    #--------------------------------------#\n",
    "    ### load Hoa et al. data (GSE135703) ###\n",
    "    hoa_t = pd.read_csv(('GSE135703/'\n",
    "                         'ExpressionMatrix.csv'),\n",
    "                         index_col=0)\n",
    "    \n",
    "    geneVersion_to_mouseName = pd.read_csv(('metaAnalysis/'\n",
    "                                            'refs/geneVersion'\n",
    "                                            '_to_mouseName.csv'),\n",
    "                                            index_col=1)\n",
    "    hoa = hoa_t.T\n",
    "    hoa = hoa.loc[:,[x for x in hoa.columns.values if x in \n",
    "                     geneVersion_to_mouseName.index.values]]\n",
    "    \n",
    "    hoa.columns = [geneVersion_to_mouseName.loc[x,'Gene name'].values[0] if\n",
    "                   type(geneVersion_to_mouseName.loc[x,'Gene name'])!=str\n",
    "                   else geneVersion_to_mouseName.loc[x,'Gene name'] for x in\n",
    "                   hoa.columns.values]\n",
    "    #--------------------------------#\n",
    "    \n",
    "    \n",
    "    \n",
    "    #-----------------------------------#\n",
    "    ### compile support cell metadata ###\n",
    "    print('compiling metadata...')\n",
    "    allCells = np.concatenate((labeled_data.index.values,\n",
    "                               burns.index.values,\n",
    "                               hoa.index.values))\n",
    "    \n",
    "    metadata = pd.DataFrame(index=allCells,\n",
    "                            columns=['dataset',\n",
    "                                     'species',\n",
    "                                     'organ',\n",
    "                                     'type',\n",
    "                                     'subtype'])\n",
    "    \n",
    "    for idx in tqdm(metadata.index.values):\n",
    "    \n",
    "        # lush et al\n",
    "        if idx in labeled_data.index.values:\n",
    "            \n",
    "            metadata.loc[idx,'dataset'] = 'lush et al'\n",
    "            metadata.loc[idx,'species'] = 'D. rerio'\n",
    "            metadata.loc[idx,'organ'] = 'neuromast'\n",
    "            node = labeled_data.loc[idx,'node']\n",
    "            \n",
    "            if node in [7,8,9,14]:\n",
    "                metadata.loc[idx,'type'] = 'sc'\n",
    "                metadata.loc[idx,'subtype'] = 'central sc'\n",
    "                \n",
    "            elif node == 4:\n",
    "                metadata.loc[idx,'type'] = 'sc'\n",
    "                metadata.loc[idx,'subtype'] = 'differentiating sc'\n",
    "                \n",
    "            elif node in [3,10,11]:\n",
    "                metadata.loc[idx,'type'] = 'sc'\n",
    "                metadata.loc[idx,'subtype'] = 'D/V amplifying sc'\n",
    "                \n",
    "            elif node == 13:\n",
    "                metadata.loc[idx,'type'] = 'sc'\n",
    "                metadata.loc[idx,'subtype'] = 'A/P pole sc'\n",
    "                \n",
    "            elif node in [5,6]:\n",
    "                metadata.loc[idx,'type'] = 'mc'\n",
    "            \n",
    "            \n",
    "        # burns et al\n",
    "        if idx in burns.index.values:\n",
    "            \n",
    "            metadata.loc[idx,'dataset'] = 'burns et al'\n",
    "            metadata.loc[idx,'species'] = 'M. musculus'\n",
    "            \n",
    "            if idx in burns_phenoData_utric.index.values:\n",
    "                \n",
    "                metadata.loc[idx,'organ'] = 'utricle'\n",
    "                GroupID = burns_phenoData_utric.loc[idx,'GroupID']\n",
    "                \n",
    "                if GroupID == 'TEC':\n",
    "                    metadata.loc[idx,'type'] = 'tec'\n",
    "                    \n",
    "                elif GroupID == 'SC (i)':\n",
    "                    metadata.loc[idx,'type'] = 'sc'\n",
    "                    metadata.loc[idx,'subtype'] = 'young sc'\n",
    "                    \n",
    "                elif GroupID == 'SC (ii)':\n",
    "                    metadata.loc[idx,'type'] = 'sc'\n",
    "                    metadata.loc[idx,'subtype'] = 'mature sc'\n",
    "                    \n",
    "            elif idx in burns_phenoData_cochl1.index.values:\n",
    "                \n",
    "                metadata.loc[idx,'organ'] = 'corti'\n",
    "                GroupID = burns_phenoData_cochl1.loc[idx,'GroupID']\n",
    "                \n",
    "                if GroupID in ['NSC (i)','NSC (ii)']:\n",
    "                    metadata.loc[idx,'type'] = 'nsc'\n",
    "                    \n",
    "                elif GroupID == 'SC':\n",
    "                    metadata.loc[idx,'type'] = 'sc'\n",
    "                    \n",
    "            elif idx in burns_phenoData_cochl2.index.values:\n",
    "                \n",
    "                metadata.loc[idx,'organ'] = 'corti'\n",
    "                GroupID = burns_phenoData_cochl2.loc[idx,'GroupID']\n",
    "                \n",
    "                if GroupID == 'MedSC':\n",
    "                    metadata.loc[idx,'type'] = 'sc'\n",
    "                    metadata.loc[idx,'subtype'] = 'medial sc'\n",
    "                    \n",
    "                elif GroupID in ['LatSCa','LatSCb']:\n",
    "                    metadata.loc[idx,'type'] = 'sc'\n",
    "                    metadata.loc[idx,'subtype'] = 'lateral sc'\n",
    "    \n",
    "    \n",
    "        # hoa et al\n",
    "        elif idx in hoa.index.values:\n",
    "            \n",
    "            metadata.loc[idx,'dataset'] = 'hoa et al'\n",
    "            metadata.loc[idx,'species'] = 'M. musculus'\n",
    "            metadata.loc[idx,'organ'] = 'corti'\n",
    "            metadata.loc[idx,'type'] = 'sc'\n",
    "    \n",
    "    \n",
    "    # remove empty rows and save metadata \n",
    "    metadata = metadata.loc[metadata.index.values[~metadata['type'].isnull().values],:] \n",
    "    \n",
    "    metadata.to_csv(('metaAnalysis/'\n",
    "                     'datasets/'\n",
    "                     'metadata.csv'))\n",
    "    #-------------------------------#\n",
    "\n",
    "    \n",
    "    \n",
    "    #------------------------------------------#\n",
    "    ### common genes for dataset integration ###                \n",
    "    shared_genes = np.array(list(set(labeled_data.columns.values).intersection(\n",
    "                            set(burns.columns.values),\n",
    "                            set(hoa.columns.values))))\n",
    "    #------------------------------------------------#\n",
    "    \n",
    "    \n",
    "    \n",
    "    #--------------------------#\n",
    "    ### write zebrafish data ###\n",
    "    lush = labeled_data.loc[[idx for idx in labeled_data.index.values if\n",
    "                             idx in metadata.index.values],shared_genes].T\n",
    "    lush = lush.loc[~lush.index.duplicated(keep='first')]\n",
    "    \n",
    "    lush.to_csv(('metaAnalysis/'\n",
    "                 'datasets/'\n",
    "                 'lush.csv'))\n",
    "    #-----------------------#\n",
    "    \n",
    "    \n",
    "    \n",
    "    #----------------------#\n",
    "    ### write mouse data ###\n",
    "    burns = burns.loc[[idx for idx in burns.index.values if \n",
    "                       idx in metadata.index.values],shared_genes].T\n",
    "    burns = burns.loc[~burns.index.duplicated(keep='first')]\n",
    "    \n",
    "    burns.to_csv(('metaAnalysis/'\n",
    "                  'datasets/'\n",
    "                  'burns.csv'))\n",
    "    \n",
    "    hoa = hoa.loc[[idx for idx in hoa.index.values if \n",
    "                   idx in metadata.index.values],shared_genes].T\n",
    "    hoa = hoa.loc[~hoa.index.duplicated(keep='first')]\n",
    "    \n",
    "    hoa.to_csv(('metaAnalysis/'\n",
    "                'datasets/'\n",
    "                'hoa.csv'))\n",
    "    #---------------------#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----------------------------------------#\n",
    "########## UMAP before alignment ##########\n",
    "#-----------------------------------------#\n",
    "\n",
    "\n",
    "#-----------------------------#\n",
    "### additional dependencies ###\n",
    "from matplotlib.legend_handler import HandlerTuple\n",
    "from sklearn.decomposition import PCA\n",
    "import umap.umap_ as umap\n",
    "#-----------------------#\n",
    "\n",
    "try:\n",
    "    umap_before = pd.read_csv(('metaAnalysis/'\n",
    "                               'embeddings/'\n",
    "                               'umap_before.csv'),\n",
    "                               index_col=0) \n",
    "except:\n",
    "    \n",
    "    #----------#\n",
    "    ### UMAP ###\n",
    "    n_pcs = 100\n",
    "    reducer = umap.UMAP()\n",
    "    \n",
    "    lush = scprep.normalize.library_size_normalize(lush)\n",
    "    lush = scprep.transform.log(lush)\n",
    "    \n",
    "    burns = scprep.normalize.library_size_normalize(burns)\n",
    "    burns = scprep.transform.log(burns)\n",
    "    \n",
    "    hoa = scprep.normalize.library_size_normalize(hoa)\n",
    "    hoa = scprep.transform.log(hoa)\n",
    "    \n",
    "    data_before = np.hstack((lush.values,\n",
    "                             burns.values,\n",
    "                             hoa.values)).T\n",
    "    \n",
    "    pca_before = PCA(n_components=n_pcs\n",
    "                     ).fit_transform(data_before)\n",
    "    \n",
    "    umap_before = reducer.fit_transform(pca_before)\n",
    "    umap_before = pd.DataFrame(umap_before,columns=['UMAP_1',\n",
    "                                                    'UMAP_2'],\n",
    "                               index=np.concatenate(\n",
    "                                   (lush.columns.values,\n",
    "                                    burns.columns.values,\n",
    "                                    hoa.columns.values)))\n",
    "    \n",
    "    umap_before.to_csv(('metaAnalysis/'\n",
    "                        'embeddings/'\n",
    "                        'umap_before.csv'))\n",
    "    #-------------------------------------#\n",
    "\n",
    "\n",
    "    \n",
    "#-------------------------#\n",
    "### dataset identifiers ###\n",
    "lush_cells = metadata.index.values[np.where((metadata['dataset']=='lush et al'))]\n",
    "\n",
    "burns_cells = metadata.index.values[np.where((metadata['dataset']=='burns et al'))]\n",
    "\n",
    "hoa_cells = metadata.index.values[np.where((metadata['dataset']=='hoa et al'))]\n",
    "#-----------------------------------------------------------------------------#\n",
    "\n",
    "\n",
    "\n",
    "#---------------#\n",
    "### plot UMAP ###\n",
    "sz = 2\n",
    "lw = sz*0.1\n",
    "plt.figure(figsize=(2.5,1.25))\n",
    "\n",
    "lush_plt = plt.scatter(umap_before.loc[lush_cells,'UMAP_1'],\n",
    "                       umap_before.loc[lush_cells,'UMAP_2'],\n",
    "                       c='seagreen',\n",
    "                       edgecolor='k',\n",
    "                       linewidth=lw,\n",
    "                       s=sz)\n",
    "\n",
    "burns_plt = plt.scatter(umap_before.loc[burns_cells,'UMAP_1'],\n",
    "                        umap_before.loc[burns_cells,'UMAP_2'],\n",
    "                        c='indianred',\n",
    "                        edgecolor='k',\n",
    "                        linewidth=lw,\n",
    "                        s=sz)\n",
    "\n",
    "hoa_plt = plt.scatter(umap_before.loc[hoa_cells,'UMAP_1'],\n",
    "                      umap_before.loc[hoa_cells,'UMAP_2'],\n",
    "                      c='goldenrod',\n",
    "                      edgecolor='k',\n",
    "                      linewidth=lw,\n",
    "                      s=sz)\n",
    "\n",
    "ax = plt.gca()\n",
    "ax.axis('off')\n",
    "\n",
    "lgd = ax.legend([(lush_plt),(burns_plt,hoa_plt)],\n",
    "                [('D. rerio neuromast support \\n'\n",
    "                  '& mantle cells (lush et al.)'),\n",
    "                 ('M. mus. inner ear support \\n'\n",
    "                  'cells (multiple studies)')],\n",
    "                 loc='lower right',\n",
    "                 borderaxespad=0,\n",
    "                 bbox_to_anchor=(1,0),\n",
    "                 fontsize=4,\n",
    "                 frameon=False,\n",
    "                 markerscale=2,\n",
    "                 handler_map=\n",
    "                {tuple:HandlerTuple(\n",
    "                    ndivide=None)})\n",
    "\n",
    "for t in lgd.get_texts():\n",
    "    t.set_ha('center')\n",
    "    t.set_position((400,0))\n",
    "\n",
    "    \n",
    "# axes annotation\n",
    "plt.annotate(s='',xy=(0,0.3),\n",
    "             xytext=(0.15,0),\n",
    "             xycoords=\n",
    "             'axes fraction',\n",
    "             arrowprops=\n",
    "             dict(arrowstyle=('<->,'\n",
    "             'head_width=0.05,'\n",
    "             'head_length=0.1'),\n",
    "             connectionstyle=\n",
    "             ('angle,rad=0,'\n",
    "              'angleA=0,'\n",
    "              'angleB=-90'),\n",
    "             color='k',\n",
    "             linewidth=0.5))\n",
    "\n",
    "plt.text(s='UMAP1',\n",
    "         x=0.065,\n",
    "         y=-0.05,\n",
    "         va='center',\n",
    "         ha='center',\n",
    "         color='k',\n",
    "         fontsize=4,\n",
    "         transform=\n",
    "         ax.transAxes)\n",
    "\n",
    "plt.text(s='UMAP2',\n",
    "         x=-0.025,\n",
    "         y=0.13,\n",
    "         rotation=90,\n",
    "         va='center',\n",
    "         ha='center',\n",
    "         color='k',\n",
    "         fontsize=4,\n",
    "         transform=\n",
    "         ax.transAxes)\n",
    "\n",
    "plt.savefig(('metaAnalysis/'\n",
    "             'figures/'\n",
    "             'before.png'),\n",
    "             dpi=dpi)\n",
    "#-------------------#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#----------------------------------------#\n",
    "########## UMAP after alignment ##########\n",
    "#----------------------------------------#\n",
    "\n",
    "\n",
    "#------------------------------#\n",
    "### load aligned data from R ###\n",
    "data_after = pd.read_csv(('metaAnalysis/'\n",
    "                          'datasets/'\n",
    "                          'aligned.csv'),\n",
    "                          index_col=0).T\n",
    "\n",
    "data_after['ids'] = [x.replace(\".\",\"-\") for x in \n",
    "                     data_after.index.values]\n",
    "\n",
    "data_after.set_index('ids',\n",
    "                inplace=True)\n",
    "\n",
    "alignment_fts = pd.read_csv(('metaAnalysis/'\n",
    "                             'datasets/'\n",
    "                             'features.csv'),\n",
    "                             index_col=0)\n",
    "#---------------------------------------#\n",
    "\n",
    "\n",
    "try:\n",
    "    umap_after = pd.read_csv(('metaAnalysis/'\n",
    "                              'embeddings/'\n",
    "                              'umap_after.csv'),\n",
    "                              index_col=0) \n",
    "    \n",
    "except:\n",
    "    #----------#\n",
    "    ### UMAP ###\n",
    "    n_pcs = 100\n",
    "    reducer = umap.UMAP()\n",
    "    \n",
    "    pca_after = PCA(n_components=n_pcs).fit_transform(\n",
    "        data_after.loc[:,alignment_fts['anchors@anchor.features'].values])\n",
    "    \n",
    "    umap_after = reducer.fit_transform(pca_after)\n",
    "    umap_after = pd.DataFrame(umap_after,\n",
    "                              columns=['UMAP_1','UMAP_2'],\n",
    "                              index=data_after.index.values)\n",
    "    \n",
    "    umap_after.to_csv(('metaAnalysis/'\n",
    "                       'embeddings/'\n",
    "                       'umap_after.csv'))\n",
    "    #-----------------------------------#\n",
    "\n",
    "\n",
    "    \n",
    "#---------------#    \n",
    "### plot UMAP ###\n",
    "sz = 4\n",
    "lw = sz*0.1\n",
    "plt.figure(figsize=(1.25,1.25))\n",
    "\n",
    "lush_plt = plt.scatter(umap_after.loc[lush_cells,'UMAP_1'],\n",
    "                       umap_after.loc[lush_cells,'UMAP_2'],\n",
    "                       c='seagreen',\n",
    "                       edgecolor='k',\n",
    "                       linewidth=lw,\n",
    "                       s=sz)\n",
    "\n",
    "burns_plt = plt.scatter(umap_after.loc[burns_cells,'UMAP_1'],\n",
    "                        umap_after.loc[burns_cells,'UMAP_2'],\n",
    "                        c='indianred',\n",
    "                        edgecolor='k',\n",
    "                        linewidth=lw,\n",
    "                        s=sz)\n",
    "\n",
    "hoa_plt = plt.scatter(umap_after.loc[hoa_cells,'UMAP_1'],\n",
    "                      umap_after.loc[hoa_cells,'UMAP_2'],\n",
    "                      c='goldenrod',\n",
    "                      edgecolor='k',\n",
    "                      linewidth=lw,\n",
    "                      s=sz)\n",
    "\n",
    "ax = plt.gca()\n",
    "ax.axis('off')\n",
    "\n",
    "\n",
    "# axes annotation\n",
    "plt.annotate(s='',xy=(0,0.3),\n",
    "             xytext=(0.3,0),\n",
    "             xycoords=\n",
    "             'axes fraction',\n",
    "             arrowprops=\n",
    "             dict(arrowstyle=('<->,'\n",
    "             'head_width=0.05,'\n",
    "             'head_length=0.1'),\n",
    "             connectionstyle=\n",
    "             ('angle,rad=0,'\n",
    "              'angleA=0,'\n",
    "              'angleB=-90'),\n",
    "             color='k',\n",
    "             linewidth=0.5))\n",
    "\n",
    "plt.text(s='UMAP1',\n",
    "         x=0.13,\n",
    "         y=-0.05,\n",
    "         va='center',\n",
    "         ha='center',\n",
    "         color='k',\n",
    "         fontsize=4,\n",
    "         transform=\n",
    "         ax.transAxes)\n",
    "\n",
    "plt.text(s='UMAP2',\n",
    "         x=-0.05,\n",
    "         y=0.13,\n",
    "         rotation=90,\n",
    "         va='center',\n",
    "         ha='center',\n",
    "         color='k',\n",
    "         fontsize=4,\n",
    "         transform=\n",
    "         ax.transAxes)\n",
    "\n",
    "plt.savefig(('metaAnalysis/'\n",
    "             'figures/'\n",
    "             'after.png'),\n",
    "             dpi=dpi)\n",
    "#-------------------#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#--------------------------------------------------------#\n",
    "########## find Gaussian mixture for clustering ##########\n",
    "#--------------------------------------------------------#\n",
    "\n",
    "#-----------------------------#\n",
    "### additional dependencies ###\n",
    "from sklearn import mixture\n",
    "from kneed import KneeLocator\n",
    "#---------------------------#\n",
    "\n",
    "cluster_n = np.arange(1,21)\n",
    "sample_m = np.arange(0,100)\n",
    "\n",
    "try:\n",
    "    bic = pd.read_csv(('metaAnalysis/'\n",
    "                       'results/'\n",
    "                       'bic.csv'),\n",
    "                       index_col=0)\n",
    "    \n",
    "    aic = pd.read_csv(('metaAnalysis/'\n",
    "                       'results/'\n",
    "                       'aic.csv'),\n",
    "                       index_col=0)\n",
    "    \n",
    "except:\n",
    "\n",
    "    #--------------------------------------------------#\n",
    "    ### calculate bic & aic for different n clusters ###\n",
    "    bic = pd.DataFrame(index=sample_m,\n",
    "                       columns=cluster_n)\n",
    "\n",
    "    aic = pd.DataFrame(index=sample_m,\n",
    "                       columns=cluster_n)\n",
    "\n",
    "    for m in tqdm(sample_m):\n",
    "    \n",
    "        for n in cluster_n:\n",
    "    \n",
    "            gm = mixture.GaussianMixture(n_components=n,\n",
    "                                         covariance_type='full'\n",
    "                                         ).fit(umap_after)\n",
    "        \n",
    "            bic.loc[m,n] = gm.bic(umap_after)\n",
    "            aic.loc[m,n] = gm.aic(umap_after)\n",
    "    \n",
    "        # normalize bic & aic for each sampling\n",
    "        bic.loc[m,:] = (bic.loc[m,:]-bic.loc[m,:].min())/(bic.loc[m,:].max()-bic.loc[m,:].min())\n",
    "        aic.loc[m,:] = (aic.loc[m,:]-aic.loc[m,:].min())/(aic.loc[m,:].max()-aic.loc[m,:].min())\n",
    "        \n",
    "        \n",
    "    bic.to_csv(('metaAnalysis/'\n",
    "                'results/'\n",
    "                'bic.csv'))\n",
    "    \n",
    "    aic.to_csv(('metaAnalysis/'\n",
    "                'results/'\n",
    "                'aic.csv'))\n",
    "    #---------------------#\n",
    "\n",
    "\n",
    "\n",
    "#-------------------------#\n",
    "### determine the knees ###\n",
    "bic_knee = KneeLocator(cluster_n,\n",
    "                       bic.mean(axis=0),\n",
    "                       S=1.0,\n",
    "                       curve=\n",
    "                       'convex',\n",
    "                       direction=\n",
    "                       'decreasing')\n",
    "\n",
    "aic_knee = KneeLocator(cluster_n,\n",
    "                       aic.mean(axis=0),\n",
    "                       S=1.0,\n",
    "                       curve=\n",
    "                       'convex',\n",
    "                       direction=\n",
    "                       'decreasing')\n",
    "\n",
    "print('Detected knee at: '+\n",
    "      str(bic_knee.knee)+\n",
    "      ' and '+str(aic_knee.knee))\n",
    "#-------------------------------#\n",
    "\n",
    "\n",
    "\n",
    "#--------------------#\n",
    "### plot bic & aic ###\n",
    "plt.figure(figsize=(2.5,1.25))\n",
    "\n",
    "plt.plot(cluster_n,\n",
    "         bic.mean(axis=0),\n",
    "         linewidth=1,\n",
    "         marker='.',\n",
    "         markersize=4,\n",
    "         c='tab:blue',\n",
    "         label='BIC')\n",
    "\n",
    "plt.plot(cluster_n,\n",
    "         aic.mean(axis=0),\n",
    "         linewidth=1,\n",
    "         marker='.',\n",
    "         markersize=4,\n",
    "         c='tab:red',\n",
    "         label='AIC')\n",
    "\n",
    "plt.legend(fontsize=5,\n",
    "           frameon=False)\n",
    "\n",
    "ax = plt.gca()\n",
    "b,t = ax.get_ylim()\n",
    "\n",
    "knee = 3\n",
    "print('Knee plotted at: '+str(knee))\n",
    "plt.vlines(knee,b,t,\n",
    "           color='k',\n",
    "           linestyle=\n",
    "           'dotted',\n",
    "           linewidth=1)\n",
    "\n",
    "ax.set_ylim(b,t)\n",
    "plt.xticks(cluster_n[::2])\n",
    "\n",
    "ax.set_xlabel('Clusters',\n",
    "              fontsize=5)\n",
    "\n",
    "ax.set_ylabel('Normalized Score',\n",
    "              fontsize=5)\n",
    "\n",
    "ax.tick_params(axis='both',\n",
    "               which='major',\n",
    "               labelsize=5)\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(('metaAnalysis/'\n",
    "             'figures/'\n",
    "             'crit.png'),\n",
    "             dpi=dpi)\n",
    "#-------------------#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----------------------------------#\n",
    "########## plot by cluster ##########\n",
    "#-----------------------------------#\n",
    "\n",
    "\n",
    "try:\n",
    "    labels_df = pd.read_csv(('metaAnalysis/'\n",
    "                             'embeddings/'\n",
    "                             'labels.csv'),\n",
    "                             index_col=0)\n",
    "    \n",
    "except:\n",
    "    \n",
    "    #----------------------------------------------------#\n",
    "    ### compute Gaussians and assign cells to clusters ###\n",
    "    gm = mixture.GaussianMixture(n_components=knee,\n",
    "                                 covariance_type='full'\n",
    "                                 ).fit(umap_after)\n",
    "\n",
    "    labels_df = pd.DataFrame(gm.predict(umap_after),\n",
    "                             index=umap_after.index.values,\n",
    "                             columns=['gmm_cluster'])\n",
    "    \n",
    "    labels_df.to_csv(('metaAnalysis/'\n",
    "                      'embeddings/'\n",
    "                      'labels.csv'))\n",
    "    #------------------------------#\n",
    "    \n",
    "    \n",
    "    \n",
    "#-------------------------------#\n",
    "### identifiers by cluster ###\n",
    "clus0 = metadata.index.values[np.where((labels_df['gmm_cluster']==0))]\n",
    "clus1 = metadata.index.values[np.where((labels_df['gmm_cluster']==1))]\n",
    "clus2 = metadata.index.values[np.where((labels_df['gmm_cluster']==2))]\n",
    "#--------------------------------------------------------------------#\n",
    "\n",
    "\n",
    "\n",
    "#-----------------#\n",
    "### plot figure ###\n",
    "sz = 4\n",
    "lw = sz*0.1\n",
    "plt.figure(figsize=(1.25,1.25))\n",
    "\n",
    "plt.scatter(umap_after.loc[clus0,'UMAP_1'],\n",
    "            umap_after.loc[clus0,'UMAP_2'],\n",
    "            label='0',\n",
    "            c='slateblue',\n",
    "            edgecolor='k',\n",
    "            linewidth=lw,\n",
    "            s=sz)\n",
    "\n",
    "plt.scatter(umap_after.loc[clus1,'UMAP_1'],\n",
    "            umap_after.loc[clus1,'UMAP_2'],\n",
    "            label='1',\n",
    "            c='cadetblue',\n",
    "            edgecolor='k',\n",
    "            linewidth=lw,\n",
    "            s=sz)\n",
    "\n",
    "plt.scatter(umap_after.loc[clus2,'UMAP_1'],\n",
    "            umap_after.loc[clus2,'UMAP_2'],\n",
    "            label='2',\n",
    "            c='sandybrown',\n",
    "            edgecolor='k',\n",
    "            linewidth=lw,\n",
    "            s=sz)\n",
    "\n",
    "lgd = plt.legend(fontsize=4,\n",
    "                 frameon=False,\n",
    "                 markerscale=1.5,\n",
    "                 loc='lower right',\n",
    "                 borderaxespad=0,\n",
    "                 bbox_to_anchor=(1,0))\n",
    "\n",
    "ax = plt.gca()\n",
    "ax.axis('off')\n",
    "\n",
    "# axes annotation\n",
    "plt.annotate(s='',xy=(0,0.3),\n",
    "             xytext=(0.3,0),\n",
    "             xycoords=\n",
    "             'axes fraction',\n",
    "             arrowprops=\n",
    "             dict(arrowstyle=('<->,'\n",
    "             'head_width=0.05,'\n",
    "             'head_length=0.1'),\n",
    "             connectionstyle=\n",
    "             ('angle,rad=0,'\n",
    "              'angleA=0,'\n",
    "              'angleB=-90'),\n",
    "             color='k',\n",
    "             linewidth=0.5))\n",
    "\n",
    "plt.text(s='UMAP1',\n",
    "         x=0.13,\n",
    "         y=-0.05,\n",
    "         va='center',\n",
    "         ha='center',\n",
    "         color='k',\n",
    "         fontsize=4,\n",
    "         transform=\n",
    "         ax.transAxes)\n",
    "\n",
    "plt.text(s='UMAP2',\n",
    "         x=-0.05,\n",
    "         y=0.13,\n",
    "         rotation=90,\n",
    "         va='center',\n",
    "         ha='center',\n",
    "         color='k',\n",
    "         fontsize=4,\n",
    "         transform=\n",
    "         ax.transAxes)\n",
    "\n",
    "plt.savefig(('metaAnalysis/'\n",
    "             'figures/'\n",
    "             'cluster.png'),\n",
    "             dpi=dpi)\n",
    "#-------------------#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#--------------------------------------#\n",
    "########## plot by cell cycle ##########\n",
    "#--------------------------------------#\n",
    "\n",
    "\n",
    "#----------------------------#\n",
    "### load cell cycle labels ###\n",
    "phases = pd.read_csv(('metaAnalysis/'\n",
    "                      'datasets/'\n",
    "                      'phases.csv'),\n",
    "                      index_col=0)\n",
    "\n",
    "phases['ids'] = data_after.index.values\n",
    "\n",
    "phases.set_index('ids',\n",
    "                 inplace=True)\n",
    "\n",
    "phases.rename({phases.columns.values[0]:'phase'},\n",
    "              axis=1,\n",
    "              inplace=True)\n",
    "#-------------------------#\n",
    "\n",
    "\n",
    "\n",
    "#-------------------------------#\n",
    "### identifiers by cell cycle ###\n",
    "g1 = metadata.index.values[np.where((phases['phase']=='G1'))]\n",
    "s = metadata.index.values[np.where((phases['phase']=='S'))]\n",
    "g2m = metadata.index.values[np.where((phases['phase']=='G2M'))]\n",
    "#-------------------------------------------------------------#\n",
    "\n",
    "\n",
    "\n",
    "#-----------------#\n",
    "### plot figure ###\n",
    "sz = 4\n",
    "lw = sz*0.1\n",
    "plt.figure(figsize=(1.25,1.25))\n",
    "\n",
    "c=['slategray',\n",
    "   'royalblue',\n",
    "   'orchid']\n",
    "\n",
    "for idx in metadata.index.values:\n",
    "    \n",
    "    phase = phases.loc[idx,'phase']\n",
    "    \n",
    "    if phase == 'G1':\n",
    "\n",
    "        plt.scatter(umap_after.loc[idx,'UMAP_1'],\n",
    "            umap_after.loc[idx,'UMAP_2'],\n",
    "            c=c[0],\n",
    "            edgecolor='k',\n",
    "            linewidth=lw,\n",
    "            s=sz)\n",
    "        \n",
    "    elif phase == 'S':\n",
    "\n",
    "        plt.scatter(umap_after.loc[idx,'UMAP_1'],\n",
    "            umap_after.loc[idx,'UMAP_2'],\n",
    "            c=c[1],\n",
    "            edgecolor='k',\n",
    "            linewidth=lw,\n",
    "            s=sz)\n",
    "        \n",
    "    else:\n",
    "\n",
    "        plt.scatter(umap_after.loc[idx,'UMAP_1'],\n",
    "            umap_after.loc[idx,'UMAP_2'],\n",
    "            c=c[2],\n",
    "            edgecolor='k',\n",
    "            linewidth=lw,\n",
    "            s=sz)\n",
    "        \n",
    "        \n",
    "g1_lgd = plt.scatter([],[],\n",
    "                     c=c[0],\n",
    "                     label='G1',\n",
    "                     edgecolor='k',\n",
    "                     linewidth=lw,\n",
    "                     s=sz)\n",
    "\n",
    "s_lgd = plt.scatter([],[],\n",
    "                    c=c[1],\n",
    "                    label='S',\n",
    "                    edgecolor='k',\n",
    "                    linewidth=lw,\n",
    "                    s=sz)\n",
    "\n",
    "g2m_lgd = plt.scatter([],[],\n",
    "                      c=c[2],\n",
    "                      label='G2M',\n",
    "                      edgecolor='k',\n",
    "                      linewidth=lw,\n",
    "                      s=sz)\n",
    "\n",
    "\n",
    "lgd = plt.legend(handles=[g1_lgd,\n",
    "                          s_lgd,\n",
    "                          g2m_lgd],\n",
    "                 fontsize=4,\n",
    "                 frameon=False,\n",
    "                 markerscale=1.5,\n",
    "                 loc='lower right',\n",
    "                 borderaxespad=0,\n",
    "                 bbox_to_anchor=(1,0))\n",
    "\n",
    "for t in lgd.get_texts():\n",
    "    t.set_ha('center')\n",
    "    t.set_position((50,0))\n",
    "\n",
    "ax = plt.gca()\n",
    "ax.axis('off')\n",
    "\n",
    "# axes annotation\n",
    "plt.annotate(s='',xy=(0,0.3),\n",
    "             xytext=(0.3,0),\n",
    "             xycoords=\n",
    "             'axes fraction',\n",
    "             arrowprops=\n",
    "             dict(arrowstyle=('<->,'\n",
    "             'head_width=0.05,'\n",
    "             'head_length=0.1'),\n",
    "             connectionstyle=\n",
    "             ('angle,rad=0,'\n",
    "              'angleA=0,'\n",
    "              'angleB=-90'),\n",
    "             color='k',\n",
    "             linewidth=0.5))\n",
    "\n",
    "plt.text(s='UMAP1',\n",
    "         x=0.13,\n",
    "         y=-0.05,\n",
    "         va='center',\n",
    "         ha='center',\n",
    "         color='k',\n",
    "         fontsize=4,\n",
    "         transform=\n",
    "         ax.transAxes)\n",
    "\n",
    "plt.text(s='UMAP2',\n",
    "         x=-0.05,\n",
    "         y=0.13,\n",
    "         rotation=90,\n",
    "         va='center',\n",
    "         ha='center',\n",
    "         color='k',\n",
    "         fontsize=4,\n",
    "         transform=\n",
    "         ax.transAxes)\n",
    "\n",
    "plt.savefig(('metaAnalysis/'\n",
    "             'figures/'\n",
    "             'cycle.png'),\n",
    "             dpi=dpi)\n",
    "#-------------------#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# med_lat = data_before.loc[data_before.index.values[np.where((metadata['organ']=='corti')&\n",
    "#                          (metadata['type']=='sc')&(metadata['dataset']=='hoa et al'))],\n",
    "#                           ['Plp1','Plp2','Prox1','Cdh4','Sparcl1','Cntn1','Fabp7','Gjb2','Cdh1',\n",
    "#                            'Anxa5','Fabp7','Matn4','Emid1','Npy','Ppp1r2','Serpine2','Hes5',\n",
    "#                            'S100a1','Lfng','Egfl6','Nupr1']]\n",
    "\n",
    "# med_lat_pca = sklearn.decomposition.PCA(n_components=4).fit_transform(med_lat)\n",
    "\n",
    "# reducer = umap.UMAP()\n",
    "# med_lat_umap = reducer.fit_transform(med_lat_pca)\n",
    "\n",
    "# plt.scatter(med_lat_umap[:,0],med_lat_umap[:,1],s=5,edgecolor='k')\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #----------------------------------------------------#\n",
    "# ########## cell types in segregated cluster ##########\n",
    "# #----------------------------------------------------#\n",
    "\n",
    "\n",
    "# #---------------------------------------#\n",
    "# ### concatenate datasets (raw counts) ###\n",
    "# data_before = pd.DataFrame(np.hstack((lush.values,\n",
    "#                                       burns.values,\n",
    "#                                       hoa.values)).T,\n",
    "                           \n",
    "#                            columns=lush.index.values,\n",
    "#                            index=np.concatenate(\n",
    "#                                (lush.columns.values,\n",
    "#                                 burns.columns.values,\n",
    "#                                 hoa.columns.values)))\n",
    "# #---------------------------------------------------#\n",
    "\n",
    "\n",
    "# #------------------------------------------------#\n",
    "# ### cell type counts in the segregated cluster ###\n",
    "\n",
    "# dvsc = np.sum((labels_df['gmm_cluster']==1)&\n",
    "#               (metadata['subtype']=='D/V amplifying sc'))\n",
    "\n",
    "# diffsc = np.sum((labels_df['gmm_cluster']==1)&\n",
    "#                 (metadata['subtype']=='differentiating sc'))\n",
    "\n",
    "# iphc_bc = np.sum((labels_df['gmm_cluster']==1)&\n",
    "#                  (metadata['organ']=='corti')&\n",
    "#                  (metadata['type']=='sc')&\n",
    "#                  (data_before[['Plp1','Plp2']].sum(axis=1)>0))\n",
    "\n",
    "# fish_other = np.sum((labels_df['gmm_cluster']==1)&\n",
    "#                     (metadata['species']=='D. rerio')&\n",
    "#                     (metadata['subtype']!='D/V amplifying sc')&\n",
    "#                     (metadata['subtype']!='differentiating sc'))\n",
    "\n",
    "# mouse_other = np.sum((labels_df['gmm_cluster']==1))-\\\n",
    "#                       dvsc-diffsc-iphc_bc-fish_other     \n",
    "# #--------------------------------------------------#\n",
    "\n",
    "\n",
    "# #-----------------#\n",
    "# ### plot figure ###\n",
    "# _,(a0, a1) = plt.subplots(2,1,gridspec_kw=\n",
    "#                           {'height_ratios':[8, 1]},\n",
    "#                           figsize=(1.25,1.25))\n",
    "\n",
    "# pie = a0.pie([dvsc,\n",
    "#               diffsc,\n",
    "#               fish_other,\n",
    "#               iphc_bc,\n",
    "#               mouse_other],\n",
    "#              colors=\n",
    "#              ['dimgray',\n",
    "#               'darkgray',\n",
    "#               'lightgray',\n",
    "#               'whitesmoke',\n",
    "#               'black'])\n",
    "\n",
    "# a1.axis(\"off\") \n",
    "# a1.legend(handles=pie[0],\n",
    "#            labels=[\"D/V sc's (D. rerio)\",\n",
    "#                    \"diff. sc's (D. rerio)\",\n",
    "#                    'Other (D. rerio)',\n",
    "#                    \"IPhC's & BC's (M. mus.)\",\n",
    "#                    'Other (M. mus.)'],\n",
    "#           loc='center',\n",
    "#           frameon=False,\n",
    "#           fontsize=4)\n",
    "\n",
    "# plt.savefig(('metaAnalysis/'\n",
    "#              'figures/'\n",
    "#              'type.png'),\n",
    "#             dpi=dpi)\n",
    "# #------------------#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# IT LOOKS LIKE DVSCs IN CLUSTER 1 ARE CYCLING MORE THAN OTHER DVSCs? HOW?\n",
    "\n",
    "clus1 = metadata.index.values[np.where((labels_df['gmm_cluster']==1)&\n",
    "                                       (phases['phase']!='G1')&\n",
    "                                       (metadata['subtype']=='D/V amplifying sc'))]\n",
    "\n",
    "clus02 = metadata.index.values[np.where((labels_df['gmm_cluster']!=1)&\n",
    "                                        (phases['phase']!='G1')&\n",
    "                                        (metadata['subtype']=='D/V amplifying sc'))]\n",
    "\n",
    "ids = np.concatenate((clus1,clus02))\n",
    "\n",
    "\n",
    "dds = deseq.DESeqDataSetFromMatrix(countData=data_before.loc[ids,:].values.T+1,\n",
    "                                   colData=robjects.DataFrame({'label':robjects.StrVector(\n",
    "                                   ['clus1' if idx in clus1 else 'clus02' for idx in ids])}),\n",
    "                                   design=Formula('~ label'))\n",
    "dds = deseq.DESeq(dds)\n",
    "\n",
    "res = deseq.results(dds)\n",
    "\n",
    "res_df = pd.DataFrame(index=data_before.columns.values)\n",
    "res_df['P.Value'] = dollar(res,'pvalue')\n",
    "res_df['adj.P.Val'] = dollar(res,'padj')\n",
    "res_df['logFC'] = dollar(res,'log2FoldChange')\n",
    "\n",
    "res_df.sort_values('logFC',\n",
    "                   ascending=False,\n",
    "                   inplace=True)\n",
    "\n",
    "print(res_df.head(20))\n",
    "print(res_df.tail(20))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from matplotlib_venn import venn2, venn3\n",
    "\n",
    "#------------------\n",
    "### identifiers for hoa et al\n",
    "clus1_s = metadata.index.values[np.where((metadata['dataset']=='hoa et al')&\n",
    "                                         (labels_df['gmm_cluster']==1)&\n",
    "                                         (phases['phase']=='S'))]\n",
    "\n",
    "clus1_g2m = metadata.index.values[np.where((metadata['dataset']=='hoa et al')&\n",
    "                                           (labels_df['gmm_cluster']==1)&\n",
    "                                           (phases['phase']=='G2M'))]\n",
    "\n",
    "clus02 = metadata.index.values[np.where((metadata['dataset']=='hoa et al')&\n",
    "                                        (labels_df['gmm_cluster']!=1))]\n",
    "\n",
    "hoa_ids = np.concatenate((clus1_s,\n",
    "                          clus1_g2m,\n",
    "                          clus02))\n",
    "#-------------\n",
    "\n",
    "\n",
    "\n",
    "# # deseq2\n",
    "# dds = deseq.DESeqDataSetFromMatrix(countData=data_before.loc[hoa_ids,:].values.T+1,\n",
    "#                                    colData=robjects.DataFrame({'label':robjects.StrVector(\n",
    "#                                    ['clus1_s' if idx in clus1_s else 'clus1_g2m' if idx in \n",
    "#                                     clus1_g2m else 'clus02' for idx in hoa_ids])}),\n",
    "#                                    design=Formula('~ label'))\n",
    "# dds = deseq.DESeq(dds)\n",
    "\n",
    "# hoa_s = deseq.results(dds,contrast=robjects.StrVector(\n",
    "#                       ['label','clus02','clus1_s']))\n",
    "\n",
    "# hoa_g2m = deseq.results(dds,contrast=robjects.StrVector(\n",
    "#                         ['label','clus1_s','clus1_g2m']))\n",
    "\n",
    "# de_hoa = pd.DataFrame(index=data_before.columns.values)\n",
    "# de_hoa['s.P.Value'] = dollar(hoa_s,'pvalue')\n",
    "# de_hoa['s.adj.P.Val'] = dollar(hoa_s,'padj')\n",
    "# de_hoa['s.logFC'] = dollar(hoa_s,'log2FoldChange')\n",
    "# de_hoa['g2m.P.Value'] = dollar(hoa_g2m,'pvalue')\n",
    "# de_hoa['g2m.adj.P.Val'] = dollar(hoa_g2m,'padj')\n",
    "# de_hoa['g2m.logFC'] = dollar(hoa_g2m,'log2FoldChange')\n",
    "\n",
    "\n",
    "upreg_s = de_hoa.index.values[np.where((de_hoa['s.logFC']>1.25)&\n",
    "                                       (de_hoa['s.adj.P.Val']<0.05))]\n",
    "\n",
    "downreg_s = de_hoa.index.values[np.where((de_hoa['s.logFC']<-1.25)&\n",
    "                                         (de_hoa['s.adj.P.Val']<0.05))]\n",
    "\n",
    "upreg_g2m = de_hoa.index.values[np.where((de_hoa['g2m.logFC']>1.25)&\n",
    "                                         (de_hoa['g2m.adj.P.Val']<0.05))]\n",
    "\n",
    "downreg_g2m = de_hoa.index.values[np.where((de_hoa['g2m.logFC']<-1.25)&\n",
    "                                           (de_hoa['g2m.adj.P.Val']<0.05))]\n",
    "\n",
    "plt.figure(figsize=(2.5,2.5))\n",
    "venn3([set(upreg_s),set(upreg_g2m),set(downreg_g2m)])\n",
    "plt.show()\n",
    "\n",
    "plt.figure(figsize=(2.5,2.5))\n",
    "venn3([set(downreg_s),set(upreg_g2m),set(downreg_g2m)])\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## DE comparison (amplifying S) ##\n",
    "\n",
    "from matplotlib_venn import venn2\n",
    "\n",
    "# cell IDs\n",
    "hoa_ampS = metadata.index.values[\\\n",
    "                np.where(amplifying&\n",
    "                (phases['phase']=='S')&\n",
    "                (metadata['dataset']==\n",
    "                 'hoa et al'))]\n",
    "\n",
    "hoa_not = metadata.index.values[\\\n",
    "            np.where(~amplifying&\n",
    "            (metadata['dataset']==\n",
    "             'hoa et al'))]\n",
    "\n",
    "lush_ampS = metadata.index.values[\\\n",
    "                 np.where(amplifying&\n",
    "                 (phases['phase']=='S')&\n",
    "                 (metadata['dataset']==\n",
    "                  'lush et al'))]\n",
    "\n",
    "lush_not = metadata.index.values[\\\n",
    "             np.where(~amplifying&\n",
    "             (metadata['dataset']==\n",
    "              'lush et al'))]\n",
    "\n",
    "hoa_s = np.concatenate((hoa_ampS,\n",
    "                        hoa_not))\n",
    "\n",
    "lush_s = np.concatenate((lush_ampS,\n",
    "                         lush_not))\n",
    "\n",
    "\n",
    "try:\n",
    "    de_hoaS = pd.read_csv(('metaAnalysis/',\n",
    "                          'results/',\n",
    "                          'hoa_de_s.csv'),\n",
    "                          index_col=0)\n",
    "    \n",
    "    de_lushS = pd.read_csv(('metaAnalysis/',\n",
    "                           'results/',\n",
    "                           'lush_de_s.csv'),\n",
    "                           index_col=0)\n",
    "    \n",
    "except:\n",
    "    # hoa amp S vs ~amp (deseq2)\n",
    "    dds = deseq.DESeqDataSetFromMatrix(countData=\n",
    "                data_before.loc[hoa_s,:].values.T+1,\n",
    "                colData=robjects.DataFrame(\\\n",
    "                {'label':robjects.IntVector(\\\n",
    "                [1 if idx in hoa_ampS else 0\\\n",
    "                         for idx in hoa_s])}),\n",
    "                    design=Formula('~ label'))\n",
    "\n",
    "    dds = deseq.DESeq(dds)\n",
    "    ds_res = deseq.results(dds)\n",
    "\n",
    "    de_hoaS = pd.DataFrame(index=\n",
    "                data_before.columns.values)\n",
    "\n",
    "    de_hoaS['P.Value'] = dollar(ds_res,\n",
    "                                'pvalue')\n",
    "\n",
    "    de_hoaS['adj.P.Val'] = dollar(ds_res,\n",
    "                                  'padj')\n",
    "\n",
    "    de_hoaS['logFC'] = dollar(ds_res,\n",
    "                              'log2FoldChange')\n",
    "\n",
    "    de_hoaS.sort_values('logFC',\n",
    "                        ascending=False,\n",
    "                        inplace=True)\n",
    "\n",
    "    de_hoaS.to_csv(('metaAnalysis/',\n",
    "                    'results/',\n",
    "                    'hoa_de_s.csv'))\n",
    "\n",
    "\n",
    "\n",
    "    # lush amp S vs ~amp (deseq2)\n",
    "    dds = deseq.DESeqDataSetFromMatrix(countData=\n",
    "                data_before.loc[lush_s,:].values.T+1,\n",
    "                colData=robjects.DataFrame(\\\n",
    "                {'label':robjects.IntVector(\\\n",
    "                [1 if idx in lush_ampS else 0\\\n",
    "                         for idx in lush_s])}),\n",
    "                    design=Formula('~ label'))\n",
    "\n",
    "    dds = deseq.DESeq(dds)\n",
    "    ds_res = deseq.results(dds)\n",
    "\n",
    "    de_lushS = pd.DataFrame(index=\n",
    "                 data_before.columns.values)\n",
    "\n",
    "    de_lushS['P.Value'] = dollar(ds_res,\n",
    "                                 'pvalue')\n",
    "\n",
    "    de_lushS['adj.P.Val'] = dollar(ds_res,\n",
    "                                   'padj')\n",
    "\n",
    "    de_lushS['logFC'] = dollar(ds_res,\n",
    "                               'log2FoldChange')\n",
    "\n",
    "    de_lushS.sort_values('logFC',\n",
    "                         ascending=False,\n",
    "                         inplace=True)\n",
    "\n",
    "    de_lushS.to_csv(('metaAnalysis/',\n",
    "                     'results/',\n",
    "                     'lush_de_s.csv'))\n",
    "\n",
    "\n",
    "\n",
    "pval = 0.05\n",
    "thresh = 1.25\n",
    "\n",
    "deg_hoaS = de_hoaS.index.values[\\\n",
    "             np.where((de_hoaS['adj.P.Val']<pval)&\n",
    "             (de_hoaS['logFC'].abs()>np.log2(thresh)))]\n",
    "\n",
    "deg_lushS = de_lushS.index.values[\\\n",
    "               np.where((de_lushS['adj.P.Val']<pval)&\n",
    "               (de_lushS['logFC'].abs()>np.log2(thresh)))]\n",
    "\n",
    "\n",
    "# plot venn diagram\n",
    "plt.figure(figsize=(4,4))\n",
    "\n",
    "venn2([set(deg_hoaS),\n",
    "       set(deg_lushS)],\n",
    "      ('adult M. mus. \\n(hoa et al.)',\n",
    "       'D. rerio \\n(lush et al.)'),\n",
    "      set_colors=('teal',\n",
    "                  'palevioletred'),\n",
    "      alpha=1)\n",
    "\n",
    "plt.title('DE Genes'+\n",
    "          '\\nAmplifying S Phase')\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(('metaAnalysis/'\n",
    "             'figures/'\n",
    "             'de_s.png'),\n",
    "             dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## GO comparison (amplifying S) ##\n",
    "\n",
    "\n",
    "# enriched TF-gene co-ocurrence terms (Enrichr)\n",
    "enr = gp.enrichr(gene_list=list(deg_hoaS),\n",
    "                 background=list(data_after.columns.values),\n",
    "                 description='hoaS',organism='Mouse',\n",
    "                 gene_sets='Enrichr_Submissions_TF-Gene_Coocurrence',\n",
    "                 outdir='metaAnalysis/gseapy/hoaS/',\n",
    "                 cutoff=1,no_plot=True)\n",
    "\n",
    "enr_hoaS = pd.DataFrame(\\\n",
    "                enr.results).sort_values(\\\n",
    "                'Combined Score',ascending=False)\n",
    "\n",
    "enr_hoaS.to_csv('metaAnalysis/'+\n",
    "                'results/'+\n",
    "                'hoa_go_s.csv')\n",
    "\n",
    "\n",
    "\n",
    "enr = gp.enrichr(gene_list=list(deg_lushS),\n",
    "                 background=list(data_after.columns.values),\n",
    "                 description='lushS',organism='Mouse',\n",
    "                 gene_sets='Enrichr_Submissions_TF-Gene_Coocurrence',\n",
    "                 outdir='metaAnalysis/gseapy/lushS/',\n",
    "                 cutoff=1,no_plot=True)\n",
    "\n",
    "enr_lushS = pd.DataFrame(\\\n",
    "                enr.results).sort_values(\\\n",
    "                'Combined Score',ascending=False)\n",
    "\n",
    "enr_lushS.to_csv('metaAnalysis/'+\n",
    "                 'results/'+\n",
    "                 'lush_go_s.csv')\n",
    "\n",
    "\n",
    "# enriched terms\n",
    "hoa_cut = 30                # knee\n",
    "# plt.figure(figsize=(3,2))\n",
    "# plt.scatter(np.arange(enr_hoaS.shape[0]),\n",
    "#             enr_hoaS['Combined Score'],\n",
    "#             c='k',s=2)\n",
    "# ax = plt.gca()\n",
    "# l,r = ax.get_xlim()\n",
    "# plt.hlines(hoa_cut,l,r,\n",
    "#            color='r')\n",
    "# ax.set_xlim(l,r)\n",
    "# plt.show()\n",
    "\n",
    "hoaS_terms = enr_hoaS.loc[enr_hoaS.index.values[\\\n",
    "                np.where(enr_hoaS['Combined Score']>\n",
    "                                         hoa_cut)],:]\n",
    "\n",
    "\n",
    "lush_cut = 200              # knee\n",
    "# plt.figure(figsize=(3,2))\n",
    "# plt.scatter(np.arange(enr_lushS.shape[0]),\n",
    "#             enr_lushS['Combined Score'],\n",
    "#             c='k',s=2)\n",
    "# ax = plt.gca()\n",
    "# l,r = ax.get_xlim()\n",
    "# plt.hlines(lush_cut,l,r,\n",
    "#            color='r')\n",
    "# ax.set_xlim(l,r)\n",
    "# plt.show()\n",
    "\n",
    "lushS_terms = enr_lushS.loc[enr_lushS.index.values[\\\n",
    "                np.where(enr_lushS['Combined Score']>\n",
    "                                         lush_cut)],:]\n",
    "\n",
    "\n",
    "# plot venn diagram\n",
    "plt.figure(figsize=(4,4))\n",
    "\n",
    "venn2([set(hoaS_terms['Term']),\n",
    "       set(lushS_terms['Term'])],\n",
    "      ('adult M. mus. \\n(hoa et al.)',\n",
    "       'D. rerio \\n(lush et al.)'),\n",
    "      set_colors=('teal',\n",
    "                  'palevioletred'),\n",
    "      alpha=1)\n",
    "\n",
    "plt.title('TF-Gene Enriched GO Terms'+\n",
    "          '\\nAmplifying S Phase')\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(('metaAnalysis/'\n",
    "             'figures/'\n",
    "             'go_s.png'),\n",
    "             dpi=300)\n",
    "\n",
    "# print zebrafish only terms\n",
    "print('Zebrafish-only terms:')\n",
    "for term in lushS_terms['Term']:\n",
    "    if term not in hoaS_terms['Term'].values:\n",
    "        print(term)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## DE comparison (amplifying G2M) ##\n",
    "\n",
    "\n",
    "# cell IDs\n",
    "hoa_ampG2M = metadata.index.values[\\\n",
    "                  np.where(amplifying&\n",
    "                  (phases['phase']=='G2M')&\n",
    "                  (metadata['dataset']==\n",
    "                   'hoa et al'))]\n",
    "\n",
    "lush_ampG2M = metadata.index.values[\\\n",
    "                   np.where(amplifying&\n",
    "                   (phases['phase']=='G2M')&\n",
    "                   (metadata['dataset']==\n",
    "                    'lush et al'))]\n",
    "\n",
    "hoa_g2m = np.concatenate((hoa_ampS,\n",
    "                          hoa_ampG2M,\n",
    "                          hoa_not))\n",
    "\n",
    "lush_g2m = np.concatenate((lush_ampS,\n",
    "                           lush_ampG2M,\n",
    "                           lush_not))\n",
    "\n",
    "\n",
    "try:\n",
    "    de_hoaG2M = pd.read_csv('metaAnalysis/'+\n",
    "                            'results/'+\n",
    "                            'hoa_de_g2m.csv',\n",
    "                            index_col=0)\n",
    "    \n",
    "    de_lushG2M = pd.read_csv('metaAnalysis/'+\n",
    "                             'results/'+\n",
    "                             'lush_de_g2m.csv',\n",
    "                             index_col=0)\n",
    "    \n",
    "except:\n",
    "    # hoa amp G2M vs amp S + ~amp (deseq2)\n",
    "    dds = deseq.DESeqDataSetFromMatrix(countData=\n",
    "                data_before.loc[hoa_g2m,:].values.T+1,\n",
    "                colData=robjects.DataFrame(\\\n",
    "                {'label':robjects.IntVector(\\\n",
    "                [1 if idx in hoa_ampG2M else 0\\\n",
    "                         for idx in hoa_g2m])}),\n",
    "                    design=Formula('~ label'))\n",
    "\n",
    "    dds = deseq.DESeq(dds)\n",
    "    ds_res = deseq.results(dds)\n",
    "    \n",
    "    de_hoaG2M = pd.DataFrame(index=\n",
    "                  data_before.columns.values)\n",
    "\n",
    "    de_hoaG2M['P.Value'] = dollar(ds_res,\n",
    "                                  'pvalue')\n",
    "\n",
    "    de_hoaG2M['adj.P.Val'] = dollar(ds_res,\n",
    "                                    'padj')\n",
    "\n",
    "    de_hoaG2M['logFC'] = dollar(ds_res,\n",
    "                                'log2FoldChange')\n",
    "\n",
    "    de_hoaG2M.sort_values('logFC',\n",
    "                          ascending=False,\n",
    "                          inplace=True)\n",
    "\n",
    "    de_hoaG2M.to_csv(('metaAnalysis/',\n",
    "                      'results/',\n",
    "                      'hoa_de_g2m.csv'))\n",
    "\n",
    "\n",
    "\n",
    "    # lush amp G2M vs amp S + ~amp (deseq2)\n",
    "    dds = deseq.DESeqDataSetFromMatrix(countData=\n",
    "                data_before.loc[lush_g2m,:].values.T+1,\n",
    "                colData=robjects.DataFrame(\\\n",
    "                {'label':robjects.IntVector(\\\n",
    "                [1 if idx in lush_ampG2M else 0\\\n",
    "                         for idx in lush_g2m])}),\n",
    "                      design=Formula('~ label'))\n",
    "\n",
    "    dds = deseq.DESeq(dds)\n",
    "    ds_res = deseq.results(dds)\n",
    "    \n",
    "    de_lushG2M = pd.DataFrame(index=\n",
    "                   data_before.columns.values)\n",
    "\n",
    "    de_lushG2M['P.Value'] = dollar(ds_res,\n",
    "                                   'pvalue')\n",
    "\n",
    "    de_lushG2M['adj.P.Val'] = dollar(ds_res,\n",
    "                                     'padj')\n",
    "\n",
    "    de_lushG2M['logFC'] = dollar(ds_res,\n",
    "                                 'log2FoldChange')\n",
    "\n",
    "    de_lushG2M.sort_values('logFC',\n",
    "                           ascending=False,\n",
    "                           inplace=True)\n",
    "\n",
    "    de_lushG2M.to_csv(('metaAnalysis/',\n",
    "                       'results/',\n",
    "                       'lush_de_g2m.csv'))\n",
    "\n",
    "\n",
    "\n",
    "pval = 0.05\n",
    "thresh = 1.25\n",
    "\n",
    "deg_hoaG2M = de_hoaG2M.index.values[\\\n",
    "                np.where((de_hoaG2M['adj.P.Val']<pval)&\n",
    "                (de_hoaG2M['logFC'].abs()>np.log2(thresh)))]\n",
    "\n",
    "deg_lushG2M = de_lushG2M.index.values[\\\n",
    "                  np.where((de_lushG2M['adj.P.Val']<pval)&\n",
    "                  (de_lushG2M['logFC'].abs()>np.log2(thresh)))]\n",
    "\n",
    "\n",
    "# plot venn diagram\n",
    "plt.figure(figsize=(4,4))\n",
    "\n",
    "venn2([set(deg_hoaG2M),\n",
    "       set(deg_lushG2M)],\n",
    "      ('adult M. mus. \\n(hoa et al.)',\n",
    "       'D. rerio \\n(lush et al.)'),\n",
    "      set_colors=('teal',\n",
    "                  'palevioletred'),\n",
    "      alpha=1)\n",
    "\n",
    "plt.title('DE Genes'+\n",
    "          '\\nAmplifying G2M Phase')\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(('metaAnalysis/'\n",
    "             'figures/'\n",
    "             'de_g2m.png'),\n",
    "             dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## \n",
    "limma = importr('limma')\n",
    "\n",
    "# cell IDs\n",
    "dvsc_s = metadata.index.values[\\\n",
    "            np.where(amplifying&\n",
    "             (phases['phase']=='S')&\n",
    "             (metadata['subtype']==\n",
    "              'D/V amplifying sc'))]\n",
    "\n",
    "corti_s = metadata.index.values[\\\n",
    "            np.where(amplifying&\n",
    "             (phases['phase']=='S')&\n",
    "             (metadata['dataset']==\n",
    "                      'hoa et al')&\n",
    "             (metadata['organ']==\n",
    "                          'corti')&\n",
    "             (metadata['type']=='sc'))]\n",
    "\n",
    "# dvsc_g2m = metadata.index.values[np.where(amplifying&\n",
    "#                                  (phases['phase']=='G2M')&\n",
    "#                                  (metadata['subtype']==\n",
    "#                                   'D/V amplifying sc'))]\n",
    "\n",
    "# corti_g2m = metadata.index.values[np.where(amplifying&\n",
    "#                                   (phases['phase']=='G2M')&\n",
    "#                                   (metadata['organ']=='corti')&\n",
    "#                                   (metadata['type']=='sc'))]\n",
    "                                   \n",
    "S = np.concatenate((dvsc_s,\n",
    "                    corti_s))\n",
    "# G2M = np.concatenate((dvsc_g2m,corti_g2m))\n",
    "\n",
    "\n",
    "# amplifying S phase (limma, using aligned data not raw counts)\n",
    "data_s = data_after.loc[S,:]\n",
    "design = pd.DataFrame(index=data_s.index.values,\n",
    "                      columns=['fish',\n",
    "                               'fish vs. mouse'])\n",
    "design['fish']=1\n",
    "design['fish vs. mouse']=0\n",
    "\n",
    "design.loc[corti_s,\n",
    "           'fish vs. mouse']=1\n",
    "\n",
    "fit = limma.lmFit(data_s.values.T,\n",
    "                  design.values)\n",
    "\n",
    "fit = limma.eBayes(fit,\n",
    "                   trend=True)\n",
    "\n",
    "de_s = pd.DataFrame(limma.topTable(fit,\n",
    "                    coef=design.shape[1],\n",
    "                    number=data_s.shape[1],\n",
    "                    **{'sort.by':'none'}),\n",
    "                    index=data_s.columns.values)\n",
    "\n",
    "de_s.sort_values('logFC',\n",
    "                 ascending=False,\n",
    "                 inplace=True)\n",
    "\n",
    "# de_dvsc.to_csv('metaAnalysis/results/de_dvsc.csv')\n",
    "\n",
    "print(de_s.loc[de_s.index.values[\\\n",
    "        np.where((de_s['adj.P.Val']<0.001)&\n",
    "        (de_s['logFC'].abs()>np.log2(1.25)))],:])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## markers for proliferating fish & mouse cells ## grn inference ##\n",
    "from arboreto.algo import grnboost2\n",
    "limma = importr('limma')\n",
    "\n",
    "# mouse TFs (via TFDB3)\n",
    "mouse_tfs = pd.read_csv('metaAnalysis/refs/mouse_tfs.csv')\n",
    "mouse_tfs.loc[mouse_tfs.index.values[-1]+1,'Symbol'] = 'Cbx5'\n",
    "\n",
    "try:\n",
    "    network = pd.read_csv('metaAnalysis/results/grn.csv',index_col=0)\n",
    "    \n",
    "except:    \n",
    "    # GRN inference with GRNboost2 via Arboreto (all cells, not just near dvsc)\n",
    "    network = grnboost2(expression_data=data_after,\n",
    "                        tf_names=list(mouse_tfs['Symbol']))\n",
    "    network.set_index(np.arange(network.shape[0]),inplace=True)\n",
    "    network['weight'] = network['importance']\n",
    "    network.drop('importance',axis=1,inplace=True)\n",
    "    network.to_csv('metaAnalysis/results/grn.csv')\n",
    "    \n",
    "cut = 4 # cutoff weight for graph edges\n",
    "# plt.figure()\n",
    "# plt.plot(network.index.values,network['weight'])\n",
    "# ax = plt.gca()\n",
    "# l,r = ax.get_xlim()\n",
    "# plt.hlines(cut,l,r,color='r')\n",
    "# ax.set_xlim(l,r)\n",
    "# plt.show()\n",
    "\n",
    "# DE analysis with limma (using integrated data, not raw counts)\n",
    "# cells near dvsc, mouse vs. fish\n",
    "de_data = data_after.loc[np.concatenate((mouse_near_dvsc,dvsc_near_dvsc)),:]\n",
    "design = pd.DataFrame(index=de_data.index.values,\n",
    "                      columns=['fish','fish vs. mouse'])\n",
    "design['fish']=1\n",
    "design['fish vs. mouse']=0\n",
    "design.loc[mouse_near_dvsc,'fish vs. mouse']=1\n",
    "fit = limma.lmFit(de_data.values.T,design.values)\n",
    "fit = limma.eBayes(fit,trend=True)\n",
    "de_dvsc = pd.DataFrame(limma.topTable(fit,\n",
    "                       coef=design.shape[1],\n",
    "                       number=de_data.shape[1],\n",
    "                       **{'sort.by':'none'}),\n",
    "                       index=de_data.columns.values)\n",
    "de_dvsc.sort_values('logFC',ascending=False,inplace=True)\n",
    "de_dvsc.to_csv('metaAnalysis/results/de_dvsc.csv')\n",
    "\n",
    "# choose (in)significant genes from DE analysis\n",
    "cutoff = 1.33\n",
    "ins = de_dvsc.index.values[np.where((de_dvsc['P.Value']>0.05)|\n",
    "                                    (de_dvsc['logFC'].abs()<np.log2(cutoff)))]\n",
    "sig = de_dvsc.index.values[np.where((de_dvsc['P.Value']<0.05)&\n",
    "                                    (de_dvsc['logFC'].abs()>np.log2(cutoff)))]\n",
    "\n",
    "# find GRN for DE genes\n",
    "G = network.loc[network.index.values[np.where((network['weight']>cut)&\n",
    "                np.array([gene in sig for gene in network['TF'].values])&\n",
    "                np.array([gene in sig for gene in network['target'].values]))],:]\n",
    "    \n",
    "# volcano plot\n",
    "plt.figure(figsize=(4,6))\n",
    "plt.scatter(de_dvsc.loc[ins,'logFC'],\n",
    "            -np.log10(de_dvsc.loc[ins,'P.Value']),\n",
    "            c='gray',s=10)\n",
    "plt.scatter(de_dvsc.loc[sig,'logFC'],\n",
    "            -np.log10(de_dvsc.loc[sig,'P.Value']),\n",
    "            c='black',s=10)\n",
    "plt.scatter(de_dvsc.loc[G['TF'].unique(),'logFC'],\n",
    "            -np.log10(de_dvsc.loc[G['TF'].unique(),'P.Value']),\n",
    "            c='indianred',s=10)\n",
    "\n",
    "# draw and annotate cutoffs\n",
    "ax = plt.gca()\n",
    "l,r = ax.get_xlim()\n",
    "b,t = ax.get_ylim()\n",
    "plt.hlines(-np.log10(0.05),-max([l,r]),max([l,r]),\n",
    "           linestyle='--',linewidth=1,color='k',zorder=0)\n",
    "plt.text(s='p = 0.05',x=l-0.075,y=-np.log10(0.05),\n",
    "         va='center',ha='right',fontsize=8)\n",
    "plt.vlines(np.log2(cutoff),b,t,\n",
    "           linestyle='--',linewidth=1,color='k',zorder=0)\n",
    "plt.vlines(-np.log2(cutoff),b,t,\n",
    "           linestyle='--',linewidth=1,color='k',zorder=0)\n",
    "plt.text(s='FC = +/- 1.33',x=0,y=17,ha='center')\n",
    "plt.annotate(s='',xy=(0.22,17),xytext=(0.4,16),\n",
    "             arrowprops=dict(arrowstyle=ArrowStyle.CurveA(),color='k'))\n",
    "plt.annotate(s='',xy=(-0.22,17),xytext=(-0.4,16),\n",
    "             arrowprops=dict(arrowstyle=ArrowStyle.CurveA(),color='k'))\n",
    "ax.set_xlim(-max([l,r]),max([l,r]))\n",
    "ax.set_ylim(b,t)\n",
    "\n",
    "# # show regulation results\n",
    "# print('DE genes: ')\n",
    "# print(de_dvsc.loc[sig,:])\n",
    "# print('\\nRegulation: ')\n",
    "# print(G)\n",
    "# print('\\nRegulation stats: ')\n",
    "# print(G['target'].value_counts().to_frame().reset_index().rename(columns={'index':'target', 'target':'count'}))\n",
    "\n",
    "# annotate genes\n",
    "plt.text(x=np.mean([-max([l,r]),-np.log2(cutoff)]),\n",
    "         s='Cbx5',y=8.8,ha='center',color='indianred')\n",
    "plt.text(x=np.mean([-max([l,r]),-np.log2(cutoff)]),\n",
    "         s='Ccnd2\\nRrm1\\nRrm2\\nRpa1\\nRpa2\\nFen1\\nAhcy',\n",
    "         y=3.5,ha='center',color='royalblue')\n",
    "plt.text(x=np.mean([max([l,r]),np.log2(cutoff)]),\n",
    "         s='Tubb5',y=17.5,ha='center',color='rebeccapurple')\n",
    "plt.text(x=np.mean([max([l,r]),np.log2(cutoff)]),\n",
    "         s='Fos',y=7,ha='center',color='indianred')\n",
    "plt.text(x=np.mean([max([l,r]),np.log2(cutoff)]),\n",
    "         s='Ucp2\\n*Pdia3',y=4.8,ha='center',color='rebeccapurple')\n",
    "\n",
    "# add arrows for GRN\n",
    "plt.annotate(s='',xy=(np.mean([-max([l,r]),-np.log2(cutoff)]),8.7),\n",
    "             xytext=(np.mean([-max([l,r]),-np.log2(cutoff)]),7.6),\n",
    "             arrowprops=dict(arrowstyle=ArrowStyle.CurveA(),color='k'))\n",
    "plt.annotate(s='',xy=(np.mean([max([l,r]),np.log2(cutoff)]),6.9),\n",
    "             xytext=(np.mean([max([l,r]),np.log2(cutoff)]),5.75),\n",
    "             arrowprops=dict(arrowstyle=ArrowStyle.CurveA(),color='k'))\n",
    "plt.annotate(s='',xy=(np.mean([-max([l,r]),-np.log2(cutoff)]),9.2),\n",
    "             xytext=(np.mean([max([l,r]),np.log2(cutoff)]),7.4),\n",
    "             arrowprops=dict(arrowstyle='<->',color='k',\n",
    "             connectionstyle='arc3,rad=1'))\n",
    "\n",
    "# labels and save\n",
    "ax.set_xlabel('log2 FC')\n",
    "ax.set_ylabel('-log10 p value')\n",
    "plt.tight_layout()\n",
    "plt.savefig('metaAnalysis/figures/dvsc_volcano.png',dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## markers for proliferating medial support cells (corti) ##\n",
    "####### REDO WITH DESEQ2 -> ONLY ONE DATASET AT A TIME, RAW READS #########\n",
    "\n",
    "# medial sc IDs: proliferating (near dvsc)\n",
    "mouse_dvsc_meta = metadata.loc[mouse_near_dvsc,:]\n",
    "burns_dvsc_idxs = mouse_dvsc_meta.index.values[np.where((mouse_dvsc_meta['dataset']=='burns et al')&\n",
    "                                                        (mouse_dvsc_meta['organ']=='corti')&\n",
    "                                                        (mouse_dvsc_meta['type']=='sc'))]\n",
    "hoa_dvsc_idxs = mouse_dvsc_meta.index.values[np.where((mouse_dvsc_meta['dataset']=='hoa et al')&\n",
    "                                                      (mouse_dvsc_meta['organ']=='corti')&\n",
    "                                                      (mouse_dvsc_meta['type']=='sc'))]\n",
    "burns_dvsc_medial_idxs = burns.columns.values[np.where((burns.T.loc[burns_dvsc_idxs,['Plp1']].values > 0)|\n",
    "                                                       (burns.T.loc[burns_dvsc_idxs,['Plp2']].values > 0))[0]]\n",
    "hoa_dvsc_medial_idxs = hoa.columns.values[np.where((hoa.T.loc[hoa_dvsc_idxs,['Plp1']].values > 0)|\n",
    "                                                   (hoa.T.loc[hoa_dvsc_idxs,['Plp2']].values > 0))[0]]\n",
    "dvsc_idxs = np.concatenate((burns_dvsc_medial_idxs,hoa_dvsc_medial_idxs))\n",
    "\n",
    "# medial sc IDs: near other\n",
    "mouse_other_meta = metadata.loc[mouse_near_other,:]\n",
    "burns_other_idxs = mouse_other_meta.index.values[np.where((mouse_other_meta['dataset']=='burns et al')&\n",
    "                                                          (mouse_other_meta['organ']=='corti')&\n",
    "                                                          (mouse_other_meta['type']=='sc'))]\n",
    "hoa_other_idxs = mouse_other_meta.index.values[np.where((mouse_other_meta['dataset']=='hoa et al')&\n",
    "                                                        (mouse_other_meta['organ']=='corti')&\n",
    "                                                        (mouse_other_meta['type']=='sc'))]\n",
    "burns_other_medial_idxs = burns.columns.values[np.where((burns.T.loc[burns_other_idxs,['Plp1']].values > 0)|\n",
    "                                                        (burns.T.loc[burns_other_idxs,['Plp2']].values > 0))[0]]\n",
    "hoa_other_medial_idxs = hoa.columns.values[np.where((hoa.T.loc[hoa_other_idxs,['Plp1']].values > 0)|\n",
    "                                                    (hoa.T.loc[hoa_other_idxs,['Plp2']].values > 0))[0]]\n",
    "other_idxs = np.concatenate((burns_other_medial_idxs,hoa_other_medial_idxs))\n",
    "\n",
    "\n",
    "# DE analysis with limma (using integrated data, not raw counts)\n",
    "de_data = data_after.loc[np.concatenate((other_idxs,dvsc_idxs)),:]\n",
    "design = pd.DataFrame(index=de_data.index.values,\n",
    "                      columns=['other','other vs. dvsc'])\n",
    "design['other']=1\n",
    "design['other vs. dvsc']=0\n",
    "design.loc[dvsc_idxs,'other vs. dvsc']=1\n",
    "fit = limma.lmFit(de_data.values.T,design.values)\n",
    "fit = limma.eBayes(fit,trend=True)\n",
    "de_medial = pd.DataFrame(limma.topTable(fit,\n",
    "                         coef=design.shape[1],\n",
    "                         number=de_data.shape[1],\n",
    "                         **{'sort.by':'none'}),\n",
    "                         index=de_data.columns.values)\n",
    "de_medial.sort_values('logFC',ascending=False,inplace=True)\n",
    "de_medial.to_csv('metaAnalysis/results/de_medial.csv')\n",
    "\n",
    "# show top results\n",
    "print(de_medial.head(15))\n",
    "print(de_medial.tail(15))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
